source/fsicalmanagement/dataaccess/event_repository.d

```diff
@@ -0,0 +1,259 @@
+module fsicalmanagement.dataaccess.event_repository;
+
+import fsicalmanagement.model.event : Event;
+import poodinis : Value;
+import std.algorithm : map;
+import std.conv : to;
+import std.range.interfaces : InputRange, inputRangeObject;
+import std.typecons : Nullable, nullable;
+
+/**
+ * A repository which stores `Event`s.
+ */
+interface EventRepository
+{
+    /**
+     * Saves an event to the repository.
+     * Params:
+     * event = The `Event` to save.
+     * 
+     * Returns: The saved `Event`.
+     */
+    Event save(Event event) @safe;
+
+    /**
+     * Gets all events from the repository.
+     *
+     * Returns: An `InputRange` containing all `Event`s from the repository.
+     */
+    InputRange!Event findAll() @safe;
+
+    /**
+     * Gets an event by its id from the repository.
+     * Params:
+     * id = The id of the event to get.
+     *
+     * Returns: `Nullable!Event` containing the `Event` corresponding to
+     *          $(D_PARAM id) in the repository or `null`, if no such event
+     *          exists.
+     */
+    Nullable!Event findById(const string id) @safe;
+
+    /**
+     * Removes an event from the repository.
+     * Params:
+     * id = The id of the event to remove.
+     */
+    void deleteById(const string id) @safe;
+}
+
+/**
+ * A MongoDB based implementation of `EventRepository`.
+ */
+class MongoDBEventRepository : EventRepository
+{
+    import vibe.data.bson : deserializeBson;
+    import vibe.db.mongo.collection : MongoCollection;
+
+private:
+    @Value("events")
+    MongoCollection events;
+
+public:
+
+    /**
+     * Saves an event to the configured MongoDB collection.
+     * Params:
+     * event = The `Event` to save.
+     * 
+     * Returns: The saved `Event`.
+     */
+    Event save(Event event) @safe
+    {
+        import std.conv : ConvException;
+        import vibe.data.bson : BsonObjectID, serializeToBson;
+
+        try
+        {
+            if (!BsonObjectID.fromString(event.id).valid)
+                throw new ConvException("invalid BsonObjectID.");
+        }
+        catch (ConvException)
+        {
+            event.id = BsonObjectID.generate.to!string;
+        }
+
+        events.insert(event.serializeToBson);
+
+        return event;
+    }
+
+    /**
+     * Gets all events from the configured MongoDB collection.
+     *
+     * Returns: An `InputRange` containing all `Event`s from the configured
+     *          MongoDB collection.
+     */
+    InputRange!Event findAll() @safe
+    {
+        return events.find().map!(deserializeBson!Event).inputRangeObject;
```

Comment:
How does `deserializeBson` behave on error? Maybe we should try to catch that, especially since we would need to handle corrupted data in some way.

Reply:
`deserializeBson` throws `Exception`s on error. Currently, this is not handled at all. It is not even easily visible, when this happens, because this actually executed from within the diet template. There all exceptions are caught implicitly, but not displayed in any way.

I think you are right, we definitely need to `try` `catch` those errors. Here is my first attempt at this:
`fsicalmanagement/utility/serialization.d`:
```D
module fsicalmanagement.utility.serialization;

import std.typecons : Nullable;
import vibe.data.bson : Bson;

/**
 * Nothrow wrapper around `vibe.data.bson.desrializeBson`. Logs `Exception`s
 * thrown by `vibe.data.bson.desrializeBson` as errors.
 * Params:
 * src = The BSON to deserialize.
 *
 * Returns: `Nullable!T` containing the deserialized BSON or `null`, when
 *          `vibe.data.bson.desrializeBson` throws.
 */
Nullable!T deserializeBsonNothrow(T)(Bson src) @safe nothrow
{
    import std.traits : fullyQualifiedName;
    import std.typecons : nullable;
    import vibe.core.log : logError;
    import vibe.data.bson : deserializeBson;

    Nullable!T ret;
    try
    {
        ret = src.deserializeBson!T.nullable;
    }
    catch (Exception e)
    {
        (() @trusted{
            logError("Error while deserializing BSON %s to %s:\n%s", src,
                fullyQualifiedName!T, e);
        })();
    }
    return ret;
}
```
`fsicalmanagement.dataaccess.event_repository.findAll` would look like this:
```D
    InputRange!Event findAll() @safe
    {
        import std.algorithm.iteration : filter;
        import std.range : only, take;

        return events.find().map!(deserializeBsonNothrow!Event)
            .filter!(nullableEvent => !nullableEvent.isNull)
            .map!(nullableEvent => nullableEvent.get)
            .inputRangeObject;
    }
```
What do you think about this?

We also need to do this for the conversion of mysql rows to our structs.

Reply:
Looks fine to me.